app/src/main/kotlin/at/bitfire/davdroid/repository/DavCollectionRepository.kt

```diff
@@ -288,4 +347,34 @@ class DavCollectionRepository @Inject constructor(
     private fun getVTimeZone(tzId: String): String? =
         DateUtils.ical4jTimeZone(tzId)?.toString()
 
-}
\ No newline at end of file
+
+    /*** OBSERVERS ***/
+
+    fun addOnChangeListener(observer: OnChangeListener) = synchronized(observers) {
+        observers += observer
+    }
+
+    fun removeOnChangeListener(observer: OnChangeListener) = synchronized(observers) {
+        observers.removeAll { it == observer }
+    }
+
+    /**
+     * Notifies registered listeners about changes in the collections.
+     */
+    @AnyThread
+    private fun notifyOnChangeListeners() = synchronized(observers) {
+        observers.forEach { observer ->
+            observer.onCollectionsChanged(context)
+        }
+    }
+
+    fun interface OnChangeListener {
+        /**
+         * Will be called when collections have changed.
+         * May run in worker thread!
+         */
+        @AnyThread
+        fun onCollectionsChanged(context: Context)
```

Comment:
What is the reason for passing the context here? Can it be omitted?

Reply:
When notifiying `PushRegistrationWorker.Companion`, which would implement the `OnChangeListener`, we need the context to start/enqueue the `PushRegistrationWorker` in `PushRegistrationWorker.Companion.onCollectionsChanged()`.

Reply:
However it sounds as if the `Context` is not really specifying the interaction, so it shouldn't be passed. Rather it could be injected, for instance like this:

`DavCollectionRepository` has a list of default observers that is obtained over [Hilt (multi-binding)](https://dagger.dev/dev-guide/multibindings.html). The worker provides a Hilt module to register. In this case, Hilt can inject the Context.

Don't know whether this will actually work though, but that's what I had in mind